src/vcard/mod_vcard_api.erl

```diff
@@ -0,0 +1,281 @@
+        catch E:R:Stack ->
+            ?LOG_ERROR(#{what => vcard_sm_iq_get_failed,
+                         class => E, reason => R, stacktrace => Stack}),
+            {error, error}
+        end.
+
+transform_from_graphql(Vcard) ->
+    #xmlel{name = <<"vCard">>,
+           attrs = [{<<"xmlns">>, <<"vcard-temp">>}],
+           children = lists:foldl(fun({Name, Value}, Acc) ->
+                                      Acc ++ transform_field_and_value(Name, Value)
+                                  end, [], maps:to_list(Vcard))}.
+
+construct_xmlel(Name, Children) when is_list(Children)->
+    [#xmlel{name = Name,
+            attrs = [],
+            children = Children}].
+
+transform_field_and_value(_Name, null) ->
+    [];
+transform_field_and_value(Name, Value) when is_list(Value) ->
+    lists:foldl(fun(Element, Acc) ->
+                    Acc ++ transform_field_and_value(Name, Element)
+                end, [], Value);
+transform_field_and_value(Name, Value) when is_map(Value) ->
+    construct_xmlel(from_gql_to_xml(Name), proccess_child_map(Value));
+transform_field_and_value(Name, Value) ->
+    construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Value}]).
+
+transform_subfield_and_value(_Name, null) ->
+    [];
+transform_subfield_and_value(<<"tags">>, TagsList) ->
+    lists:foldl(fun(Tag, Acc) ->
+                    Acc ++ construct_xmlel(Tag, [])
+                end, [], TagsList);
+transform_subfield_and_value(Name, Value) when is_list(Value) ->
+    lists:foldl(fun(Element, Acc) ->
+                    Acc ++ construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Element}])
+                end, [], Value);
+transform_subfield_and_value(Name, Value) ->
+    construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Value}]).
+
+proccess_child_map(Value) ->
+    lists:foldl(fun({Name, SubfieldValue}, Acc) ->
+                    Acc ++ transform_subfield_and_value(Name, SubfieldValue)
+                end, [], maps:to_list(Value)).
+
+from_gql_to_xml(<<"formattedName">>) -> <<"FN">>;
+from_gql_to_xml(<<"nameComponents">>) -> <<"N">>;
+from_gql_to_xml(<<"birthday">>) -> <<"BDAY">>;
+from_gql_to_xml(<<"address">>) -> <<"ADR">>;
+from_gql_to_xml(<<"telephone">>) -> <<"TEL">>;
+from_gql_to_xml(<<"timeZone">>) -> <<"TZ">>;
+from_gql_to_xml(<<"sortString">>) -> <<"SOR">>;
+from_gql_to_xml(<<"givenName">>) -> <<"GIVEN">>;
+from_gql_to_xml(<<"middleName">>) -> <<"MIDDLE">>;
+from_gql_to_xml(<<"credential">>) -> <<"CRED">>;
+from_gql_to_xml(<<"country">>) -> <<"CTRY">>;
+from_gql_to_xml(Name) -> list_to_binary(string:to_upper(binary_to_list(Name))).
+
+to_gql_format(Vcard) ->
+    lists:foldl(fun(#xmlel{name = Name, children = Value}, Acc) ->
+        maps:merge(Acc, transform_from_xml(Name, Value, Acc))
+    end, #{}, Vcard).
+
+transform_from_xml(<<"FN">>, [{_, Value}], _) ->
+    #{<<"formattedName">> => Value};
+transform_from_xml(<<"N">>, Value, _) ->
+    #{<<"nameComponents">> => lists:foldl(fun name_components_proccess/2, #{}, Value)};
```

Comment:
The naming is a bit weird, I think it should be `process_something` rather than `something_process`.